vertical-pod-autoscaler/enhancements/7862-cpu-startup-boost/README.md

```diff
@@ -0,0 +1,283 @@
+  * **NEW**: `StartupBoostOnly`: new mode that will allow users to only enable
+  the startup boost feature for a container and not vanilla VPA altogether.
+  * **NEW**: `Auto`: we will modify the existing `Auto` mode to enable both
+  vanilla VPA and CPU Startup Boost (when `StartupBoost` parameter is
+  specified).
+
+#### Priority of `StartupBoost`
+
+The new `StartupBoost` field will take precedence over the rest of the container
+resource policy configurations. Functioning independently from all other fields
+in [`ContainerResourcePolicy`](https://github.com/kubernetes/autoscaler/blob/vertical-pod-autoscaler-1.3.0/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L191),
+**except for**:
+   * [`ContainerName`](https://github.com/kubernetes/autoscaler/blob/vertical-pod-autoscaler-1.3.0/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L192-L195)
+   * [`Mode`](https://github.com/kubernetes/autoscaler/blob/vertical-pod-autoscaler-1.3.0/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L196-L198)
+   * [`ControlledValues`](https://github.com/kubernetes/autoscaler/blob/vertical-pod-autoscaler-1.3.0/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L214-L217)
+
+This means that a container's CPU request/limit can be boosted during startup
+beyond [`MaxAllowed`](https://github.com/kubernetes/autoscaler/blob/vertical-pod-autoscaler-1.3.0/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L203-L206),
+for example, or it will be able to be boosted even if CPU is explicitly
+excluded from [`ControlledResources`](https://github.com/kubernetes/autoscaler/blob/vertical-pod-autoscaler-1.3.0/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1/types.go#L208-L212).
+
+### Validation
+
+#### Static Validation
+
+* We will check that the `startupBoost` configuration is valid when VPA objects
+are created/updated:
+   * The VPA autoscaling mode must be `InPlaceOrRecreate` (since it does not
+   make sense to use this feature with disruptive modes of VPA).
+   * The boost factor is >= 1 (via CRD validation rules)
+   * Only one of `StartupBoost.CPU.Factor` or `StartupBoost.CPU.Value` is
+   specified
+   * The [feature enablement](#feature-enablement) flags must be on.
+
+
+#### Dynamic Validation
+
+* `StartupBoost.CPU.Value` must be greater than the CPU request or limit of the
+  container during the boost phase, otherwise we risk downscaling the container.
+
+* Workloads must be configured with a Readiness or a Startup probe to be able to
+utilize this feature. Therefore, VPA will not boost CPU resources of workloads
+that do not configure a Readiness or a Startup probe.
+
+### Mitigating Failed In-Place Downsizes
+
+The VPA Updater **will not** evict a pod to actuate a startup CPU boost
+recommendation if it attempted to apply the recommendation in place and it
```

Comment:
I assume the process will be:
1. A new pod will start life in the boosted state
2. After the Pod has changed to Ready it will unboost

This part sounds like it's referring to the actual boosting part, which I assume is incorrect, it's the unboost part that matters here?

Reply:
Yes, it's the unboost part that matters. I rewrote this to make it clearer. Thanks.